libnavui-maneuver/src/main/java/com/mapbox/navigation/ui/maneuver/api/MapboxManeuverApi.kt

```diff
@@ -1,137 +1,154 @@
-     */
-    fun getManeuver(
-        bannerInstruction: BannerInstructions,
-        callback: ManeuverCallback
-    ) {
-        maneuverJob?.cancel()
-        maneuverJob = mainJobController.scope.launch {
-            val action = ManeuverAction.GetManeuver(bannerInstruction)
-            val result = processor.process(action) as GetManeuver
-            callback.onManeuver(ExpectedFactory.createValue(result.maneuver))
-        }
-    }
+    constructor(formatter: DistanceFormatter) : this(formatter, ManeuverProcessor)
 
     /**
-     * Given [RouteStepProgress] the method returns the distance remaining to finish the [LegStep]
-     * @param routeStepProgress RouteStepProgress progress object specific to the current step the user is on
-     * @param callback StepDistanceRemainingCallback contains [Expected]
+     * Given a [DirectionsRoute] the function iterates through all the [RouteLeg] in a [DirectionsRoute].
+     * For every [RouteLeg] the API iterates through all the [LegStep]. For every [LegStep] the API
+     * parses the [BannerInstructions], transforms it into [Maneuver] object and adds it to a list.
+     * Once this list is ready the API maps this list to the [RouteLeg].
+     *
+     * @param route DirectionsRoute The route associated
+     * @param routeLeg RouteLeg Specify to inform the API of the [RouteLeg] you wish to get the list of [Maneuver].
+     * If null, the API returns the list of maneuvers for the first [RouteLeg] in a [DirectionsRoute]
+     * @param callback ManeuverCallback invoked with appropriate result
      */
-    fun getStepDistanceRemaining(
-        routeStepProgress: RouteStepProgress,
-        callback: StepDistanceRemainingCallback
+    fun getManeuvers(
+        route: DirectionsRoute,
+        routeLeg: RouteLeg? = null,
+        callback: ManeuverCallback,
     ) {
-        stepDistanceRemainingJob?.cancel()
-        stepDistanceRemainingJob = mainJobController.scope.launch {
-            val action = ManeuverAction.GetStepDistanceRemaining(routeStepProgress)
-            val result = processor.process(action) as GetStepDistanceRemaining
-            callback.onStepDistanceRemaining(
-                ExpectedFactory.createValue(
-                    StepDistance(distanceFormatter, result.distanceRemaining)
+        val action = ManeuverAction.GetManeuverListWithRoute(
+            route,
+            routeLeg,
+            maneuverState,
+            distanceFormatter
+        )
+        when (val result = processor.process(action)) {
+            is ManeuverResult.GetManeuverList.Success -> {
+                val allManeuvers = result.maneuvers
+                callback.onManeuvers(ExpectedFactory.createValue(allManeuvers))
+            }
+            is ManeuverResult.GetManeuverList.Failure -> {
+                callback.onError(ExpectedFactory.createError(ManeuverError(result.error)))
+            }
+            else -> {
+                callback.onError(
```

Comment:
I would leave this as is because in all the other API's we are handling the same way. If we want to throw exceptions we should do it for every other API. @Guardiola31337 I am not sure if you would support this approach. 
Let's discuss this in a separate ticket.
cc @LukasPaczos

Reply:
If you're referring to https://github.com/mapbox/mapbox-navigation-android/pull/4482#discussion_r660950226 that's different. I agree with @LukasPaczos here this is more around these lines https://github.com/mapbox/mapbox-navigation-android/blob/cd9fc187fe2a966e81ad3b77e257532d67af5f80/libnavui-voice/src/main/java/com/mapbox/navigation/ui/voice/api/MapboxSpeechApi.kt#L87-L89 In this particular case it's a `check` for us to prevent developer errors because it's a state that shouldn't be generated / happened internally.